api/provider/aws/deploy_create.go

```diff
@@ -0,0 +1,134 @@
+package aws
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+func (d *DeployProvider) Create(req models.CreateDeployRequest) (*models.Deploy, error) {
+	deployID := generateEntityID(req.DeployName)
+	familyName := addLayer0Prefix(d.Config.Instance(), req.DeployName)
+
+	taskDefinition, err := d.renderTaskDefinition(req.DeployFile, familyName)
+	if err != nil {
+		return nil, err
+	}
+
+	taskDefinitionOutput, err := d.createTaskDefinition(taskDefinition)
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := json.Marshal(taskDefinitionOutput)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to extract deploy file: %s", err.Error())
+	}
+
+	deploy := &models.Deploy{
+		DeployName: req.DeployName,
+		DeployID:   deployID,
+		Version:    strconv.FormatInt(aws.Int64Value(taskDefinitionOutput.Revision), 10),
+		DeployFile: bytes,
+	}
+
+	taskDefinitionARN := aws.StringValue(taskDefinitionOutput.TaskDefinitionArn)
+
+	if err := d.createTags(deploy.DeployName, deploy.DeployID, deploy.Version, taskDefinitionARN); err != nil {
+		return deploy, err
+	}
+
+	return deploy, nil
+}
+
+func (d *DeployProvider) createTaskDefinition(taskDefinitionRequest *ecs.TaskDefinition) (*ecs.TaskDefinition, error) {
+
+	input := &ecs.RegisterTaskDefinitionInput{}
+	input.SetFamily(aws.StringValue(taskDefinitionRequest.Family))
+	input.SetTaskRoleArn(aws.StringValue(taskDefinitionRequest.TaskRoleArn))
+	input.SetNetworkMode(aws.StringValue(taskDefinitionRequest.NetworkMode))
+	input.SetContainerDefinitions(taskDefinitionRequest.ContainerDefinitions)
+	input.SetVolumes(taskDefinitionRequest.Volumes)
+	input.SetPlacementConstraints(taskDefinitionRequest.PlacementConstraints)
+
+	if err := input.Validate(); err != nil {
+		return nil, err
+	}
+
+	output, err := d.AWS.ECS.RegisterTaskDefinition(input)
+	if err != nil {
+		return nil, err
+	}
+
+	return output.TaskDefinition, nil
+}
+
+func (d *DeployProvider) renderTaskDefinition(body []byte, familyName string) (*ecs.TaskDefinition, error) {
+	var taskDefinition *ecs.TaskDefinition
+
+	if err := json.Unmarshal(body, &taskDefinition); err != nil {
+		return nil, fmt.Errorf("Failed to decode deploy: %s", err.Error())
+	}
+
+	taskDefinition.SetFamily(familyName)
+
+	if len(taskDefinition.ContainerDefinitions) == 0 {
+		return nil, fmt.Errorf("Deploy must have at least one container definition")
+	}
+
+	// Valid NetworkMode options are [bridge, host, none], "bridge" is most common so defaulting to that.
+	if taskDefinition.NetworkMode == nil {
```

Comment:
are we sure we want this? Shouldn't users be able to specify no network mode?

Reply:
I had some bizarre behavior during testing that I can no longer repro. As such, this was added to be explicit about setting a default network mode, and I'm fine with removing it. That said, a user can specify no network mode by specifying `none` as a param to `NetworkMode` in their task definition.

Reply:
perhaps we should set this block to `none` if it is nil and throw a comment like "hey, this needs to be explicitly set"